src/WasmParser.ts

```diff
@@ -1780,104 +1843,104 @@ export class BinaryReader {
+      case ElementSegmentType.DeclaredElemexpr:
+        mode = ElementMode.Declarative;
+        break;
+      default:
+        throw new Error(`Unsupported element segment type ${segmentType}`);
     }
-
     this.state = BinaryReaderState.BEGIN_ELEMENT_SECTION_ENTRY;
-    this.result = { index: tableIndex };
+    this.result = { mode, tableIndex };
     this._sectionEntriesLeft--;
-    this._segmentFlags = flags;
+    this._segmentType = segmentType;
     return true;
   }
   private readElementEntryBody(): boolean {
-    let funcType = Type.unspecified;
-    const pos = this._pos;
-    if (
-      this._segmentFlags &
-      (SegmentFlags.IsPassive | SegmentFlags.HasTableIndex)
-    ) {
-      if (!this.hasMoreBytes()) return false;
-      funcType = this.readVarInt7();
-    }
-    if (!this.hasVarIntBytes()) {
-      this._pos = pos;
-      return false;
-    }
-    const numElemements = this.readVarUint32();
-    const elements = new Uint32Array(numElemements);
-    for (let i = 0; i < numElemements; i++) {
-      if (this._segmentFlags & SegmentFlags.FunctionsAsElements) {
-        if (!this.hasMoreBytes()) {
-          this._pos = pos;
-          return false;
-        }
-        // Read initializer expression, which must either be null ref or func ref
-        let operator = this.readUint8();
-        if (operator == OperatorCode.ref_null) {
-          elements[i] = NULL_FUNCTION_INDEX;
-        } else if (operator == OperatorCode.ref_func) {
-          if (!this.hasVarIntBytes()) {
-            this._pos = pos;
-            return false;
-          }
-          elements[i] = this.readVarInt32();
-        } else {
-          this.error = new Error("Invalid initializer expression for element");
-          return true;
-        }
-        if (!this.hasMoreBytes()) {
-          this._pos = pos;
-          return false;
-        }
-        operator = this.readUint8();
-        if (operator != OperatorCode.end) {
-          this.error = new Error(
-            "Expected end of initializer expression for element"
-          );
-          return true;
-        }
-      } else {
-        if (!this.hasVarIntBytes()) {
-          this._pos = pos;
-          return false;
-        }
-        elements[i] = this.readVarUint32();
-      }
+    let elementType = Type.funcref;
+    switch (this._segmentType) {
+      case ElementSegmentType.PassiveExternval:
+      case ElementSegmentType.ActiveExternval:
+      case ElementSegmentType.DeclaredExternval:
+        if (!this.hasMoreBytes()) return false;
+        // We just skip the 0x00 byte, the `elemkind` byte
+        // is reserved for future versions of WebAssembly.
+        this.skipBytes(1);
+        break;
+      case ElementSegmentType.PassiveElemexpr:
+      case ElementSegmentType.ActiveElemexpr:
+      case ElementSegmentType.DeclaredElemexpr:
+        if (!this.hasMoreBytes()) return false;
+        elementType = this.readVarInt7();
+        break;
     }
```

Comment:
The default case is missing here;  I assume that the actual error would already be caught earlier when setting `this._segmentType` in this case, but I think that it's nice to have a default error handling case anyway.

Reply:
done